### PR TITLE
Account creation api

### PR DESCRIPTION
**Pull request by rmn-boiko**
## Description

Closes: https://github.com/kamu-data/kamu-node/issues/224

## Checklist before requesting a review

- [x] Unit and integration tests added
  
- [x] Compatibility:
    
  - [x] Network APIs: ✅
    
  - [x] Workspace layout and metadata: ✅
    
  - [x] Configuration: ✅
    
  - [x] Container images: ✅
- [x] Observability:
    
  - [x] Tracing / [metrics](https://github.com/kamu-data/kamu-standards/blob/master/metrics_design.md): ✅
    
  - [x] Alerts: ✅
- [x] Documentation:
    
  - [x] [Changelog](./CHANGELOG.md): ✅
    
  - [x] [Public documentation](https://github.com/kamu-data/kamu-docs/): ✅
  
- [ ] Downstream effects:
    
  - [ ] [kamu-node](https://github.com/kamu-data/kamu-node): ✅
    
  - [ ] [kamu-web-ui](https://github.com/kamu-data/kamu-web-ui): ✅
    
  - [ ] [release-queue](https://github.com/orgs/kamu-data/projects/4/views/1)

### CODE REVIEW

File: src/app/cli/src/app.rs [hunk 1]
```diff
@@ -822,6 +823,24 @@ pub fn register_config_in_catalog(
     }
     //
 
+    // Did secret key encryption configuration
+    // TODO: decide should we panic if the key is not set
+    let did_encryption_config = config
```
Comment (rmn-boiko) [inline src/app/cli/src/app.rs:828]: Extremely important!!!
Accrding to requirements we are storing encrypted did secret keys in DB for both `dataset and account` entities. Logic requires Encryption key. But now We are pushing users to always set this key for kamu-cli and kamu-node. usage. 
I want your thoughts on whether we should generate some encryption key automatically and store it for the user? Or we will force users to set it always.

Another my thought was to move storing did keys to the GQL level which will make possible to use this value optional for cli usage

File: resources/schema.gql [hunk 2]
```diff
@@ -134,6 +139,8 @@ type AccountMut {
 	Update account email
 	"""
 	updateEmail(newEmail: String!): UpdateEmailResult!
+	createAccount(accountName: String!, email: String): CreateAccountResult!
+	modifyPassword(accountName: AccountName!, password: String!): ModifyPasswordResult!
```
[the remark this answers is not in the document]
Comment (sergiimk) [reply resources/schema.gql:143]: Is `modifyPassword` an admin-only endpoint?
Please add docstrings to all new endpoints to explain their purpose.

File: src/adapter/graphql/src/mutations/account_mut.rs [hunk 3]
```diff
@@ -57,6 +70,81 @@ impl AccountMut {
...
+        let Ok(account_name) = odf::AccountName::from_str(&account_name) else {
+            return Ok(CreateAccountResult::InvalidAccountName(
+                AccountNameInvalid::default(),
+            ));
+        };
+
+        let new_email = email.unwrap_or(
+            if let Some(parent_account_parts) = self.account.email.to_string().split_once('@') {
+                format!(
+                    "{}_{}@{}",
```
Comment (sergiimk) [inline src/adapter/graphql/src/mutations/account_mut.rs:91]: Why use underscore, which most likely will produce an invalid email, when we could use `+` to direct all child account notifications to the parent?

File: src/app/cli/src/services/config/models.rs [hunk 4]
```diff
@@ -75,6 +76,9 @@ pub struct CLIConfig {
     /// Uploads configuration
     #[merge(strategy = merge_recursive)]
     pub uploads: Option<UploadsConfig>,
+
+    /// Did secret key encryption configuration
+    pub did_encryption: Option<DidSecretEncryptionConfig>,
```
Comment (sergiimk) [inline src/app/cli/src/services/config/models.rs:81]: I'd like us to have a more hierarchical configuration, grouping options into a few big sections.

In the node config we have `auth` section, where I think DID secrets encryption would fit. Perhaps let's add such a section into CLI and nest this configuration under it?

File: src/domain/accounts/domain/src/entities/predefined_accounts_config.rs [hunk 5]
```diff
@@ -98,6 +99,8 @@ pub struct AccountConfig {
     #[serde(default)]
     pub is_admin: bool,
     #[serde(default)]
+    pub can_provision_accounts: bool,
```
Comment (sergiimk) [inline src/domain/accounts/domain/src/entities/predefined_accounts_config.rs:102]: Not necessarily now, but I think it would make more sense for us to model these not as `is_x` flags but rather a list of `properties: [admin, can_provision_accounts]` and later relations. Otherwise we'll:
- get tired extending this type with all possible properties
- make system rigid - we want to allow for all kinds of custom ReBAC properties here

File: src/domain/datasets/domain/src/repos/dataset_did_secret_key_repository.rs [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+// by the Apache License, Version 2.0.
+
+use crypto_utils::DidSecretKey;
+use internal_error::InternalError;
+use thiserror::Error;
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[async_trait::async_trait]
+pub trait DatasetDidSecretKeyRepository: Send + Sync {
```
Comment (sergiimk) [inline src/domain/datasets/domain/src/repos/dataset_did_secret_key_repository.rs:17]: I'm second-guessing whether it makes sense to separate DID secret storage between accounts and datasets. Who knows what other entities we will be generating DIDs for - maybe it's better to store them all in one table. Ideally we should have a mechanism to tell by DID whether it belongs to a dataset, account or some other type. We somewhat have this with `did:odf` vs `did:key` but we could start embedding entity types into ODF DIDs too.

What do you think @rmn-boiko @zaychenko-sergei ?

File: src/utils/crypto-utils/src/entities/did_secret_encryption_config.rs [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+#[skip_serializing_none]
+#[derive(Debug, Default, Clone, Serialize, Deserialize)]
+#[serde(deny_unknown_fields, rename_all = "camelCase")]
+pub struct DidSecretEncryptionConfig {
+    /// The encryption key must be a 32-character alphanumeric string, which
+    /// includes both uppercase and lowercase Latin letters (A-Z, a-z) and
+    /// digits (0-9).
+    ///
+    /// # Example
```
Comment (sergiimk) [inline src/utils/crypto-utils/src/entities/did_secret_encryption_config.rs:27]: Please add a shell command that generates such a string - this is very helpful when deploying.

File: resources/schema.gql (hunk 2 above)
[the remark this answers is not in the document]
Comment (rmn-boiko) [reply resources/schema.gql:143]: Yeah, it's under `AdminGuard`
Added docstring

File: src/adapter/graphql/src/mutations/account_mut.rs (hunk 3 above)
Comment (rmn-boiko) [reply src/adapter/graphql/src/mutations/account_mut.rs:91]: Both `_` and `+` are allowed symbols and will be valid emails. But I used an underscore by mistake. Changed to `+`

File: src/domain/datasets/domain/src/repos/dataset_did_secret_key_repository.rs (hunk 6 above)
Comment (rmn-boiko) [reply src/domain/datasets/domain/src/repos/dataset_did_secret_key_repository.rs:17]: Yeah, I would love to have single place for all DID's. 
But want to hear your thoughts on how you can see entities in DID? You think about `did:account:<id>` and `did:dataset:<id>`? Or something more general?

File: src/utils/crypto-utils/src/entities/did_secret_encryption_config.rs (hunk 7 above)
Comment (rmn-boiko) [reply src/utils/crypto-utils/src/entities/did_secret_encryption_config.rs:27]: Added

File: src/domain/datasets/domain/src/repos/dataset_did_secret_key_repository.rs (hunk 6 above)
Comment (sergiimk) [reply src/domain/datasets/domain/src/repos/dataset_did_secret_key_repository.rs:17]: - We used the most simple form `did:key:<public key>` for accounts.
- In DimaP's wallet auth work it looks like we will adopt `did:ethr:<public key hash>` or `did:pkh` for identifying EVM accounts.
- We still only use `did:odf:<public key>` for datasets.

So we currently can non-ambiguously understand what key is for what entity.

If we need more entity types we could extend to `did:odf:<something>:<key>` or use first few bytes in key part to encode the entity type and other extra information (similarly how [IPFS CID](https://docs.ipfs.tech/concepts/content-addressing/#how-cids-are-created) is structured).

File: src/adapter/graphql/src/mutations/account_mut.rs (hunk 3 above)
Comment (sergiimk) [reply src/adapter/graphql/src/mutations/account_mut.rs]: Yeah sorry, by "invalid" I actually meant "email that doesn't exist"

File: resources/schema.gql (hunk 2 above)
[the remark this answers is not in the document]
Comment (s373r) [reply resources/schema.gql:149]: Optional: hide admin methods from GQL schema

File: src/domain/accounts/domain/src/entities/predefined_accounts_config.rs [hunk 8]
```diff
@@ -131,6 +131,17 @@ impl AccountConfig {
...
     }
 
+    pub fn set_properties(mut self, properties: Vec<AccountPropertyName>) -> Self {
+        for property in properties {
+            let existing_properties = self.properties.get_or_insert_with(Vec::new);
+
+            if !existing_properties.contains(&property) {
+                existing_properties.push(property);
+            }
+        }
```
Comment (s373r) [inline src/domain/accounts/domain/src/entities/predefined_accounts_config.rs:141]: Q: can we just do an assignment based on the account name?

```rust
self.properties = properties;
```

File: src/domain/accounts/services/src/services/account_service_impl.rs [hunk 9]
```diff
@@ -129,6 +161,63 @@ impl AccountService for AccountServiceImpl {
         self.account_repo
             .search_accounts_by_name_pattern(name_pattern, filters, pagination)
     }
+
+    async fn create_account(
+        &self,
+        account_name: &odf::AccountName,
+        email: email_utils::Email,
+        password: String,
```
Comment (s373r) [inline src/domain/accounts/services/src/services/account_service_impl.rs:169]: Q: maybe allocate a new type (`nutype`) that will not allow empty password and length less than 8?

File: src/domain/accounts/services/src/services/predefined_accounts_registrator.rs [hunk 10]
```diff
@@ -56,9 +56,13 @@ impl PredefinedAccountsRegistrator {
...
+            return Ok(());
+        };
+
+        for name in account_properties {
             self.rebac_service
-                .set_account_property(account_id, name, &value)
+                .set_account_property(account_id, *name, &boolean_property_value(true))
                 .await
                 .int_err()?;
         }
```
Comment (s373r) [inline src/domain/accounts/services/src/services/predefined_accounts_registrator.rs:68]: I think this solution may be fragile, and it is better to explicitly set the values. We have default values in case something is not explicitly specified

File: src/infra/accounts/repo-tests/src/accounts_repository_test_utils.rs [hunk 11]
```diff
@@ -21,9 +21,10 @@ pub(crate) fn make_test_account(
     email: &str,
     provider: &str,
     provider_identity_key: &str,
+    account_id_maybe: Option<odf::AccountID>,
```
Comment (s373r) [inline src/infra/accounts/repo-tests/src/accounts_repository_test_utils.rs:24]: Q: why do we need this parameter if generating from the name always gives a stable result?

File: src/infra/accounts/repo-tests/src/password_hash_repository_test_suite.rs [hunk 12]
```diff
@@ -76,3 +96,72 @@ pub async fn test_store_couple_account_passwords(catalog: &Catalog) {
 }
 
 ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+pub async fn test_modify_password(catalog: &Catalog) {
```
Comment (s373r) [inline src/infra/accounts/repo-tests/src/password_hash_repository_test_suite.rs:100]: Please add a test with an attempt to set an empty password

Comment (s373r) [reply src/infra/accounts/repo-tests/src/password_hash_repository_test_suite.rs:100]: PS. The same applies to the password hash

File: src/utils/crypto-utils/Cargo.toml [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+
+[lib]
+doctest = false
+
+
+[dependencies]
+aes-gcm = { version = "0.10.3" }
+argon2 = { version = "0.5" }
+internal-error = { workspace = true }
+odf = { workspace = true }
```
Comment (s373r) [inline src/utils/crypto-utils/Cargo.toml:28]: Please extract to own group

File: src/utils/crypto-utils/src/aes_gcm.rs [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+#[derive(Debug)]
+struct AesGcmError(aes_gcm::Error);
+
+impl std::fmt::Display for AesGcmError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "AES-GCM error")
+    }
+}
+
+impl std::error::Error for AesGcmError {}
```
Comment (s373r) [inline src/utils/crypto-utils/src/aes_gcm.rs:69]: Q: Why not just `thiserror::Error`?

File: src/utils/crypto-utils/src/argon2_hash.rs [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use aes_gcm::aead::OsRng;
+use argon2::PasswordHash;
+use password_hash::{PasswordHasher, PasswordVerifier, SaltString};
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+pub fn get_argon2_hash(password: &str, hashing_mode: PasswordHashingMode) -> String {
```
Comment (s373r) [inline src/utils/crypto-utils/src/argon2_hash.rs:16]: Optional: group by struct (e.g., `Argon2Hasher`) with static methods (`hash()`, `verify()`)

File: src/utils/crypto-utils/src/lib.rs [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+}
+
+impl From<ParseEncryptionKey> for EncryptionError {
+    fn from(value: ParseEncryptionKey) -> Self {
+        match value {
+            ParseEncryptionKey::InvalidEncryptionKeyLength => Self::InvalidEncryptionKey,
+            ParseEncryptionKey::InternalError(err) => Self::InternalError(err),
+        }
+    }
+}
```
Comment (s373r) [inline src/utils/crypto-utils/src/lib.rs:56]: Please extract to own file

File: src/utils/random-strings/src/common.rs [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use rand::distributions::{Alphanumeric, Uniform};
+use rand::prelude::Distribution;
+
+pub fn get_random_string(
```
Comment (s373r) [inline src/utils/random-strings/src/common.rs:13]: I would like to suggest splitting this method into two:
- `get_random_name()` -- keep the old 
- `get_random_string()` -- add a new

`get_random_name()`, based on where it is used, was intended to return readable names (as `Docker` does, for example)

**Review by s373r — COMMENTED** [review]
Added comments/questions

File: src/infra/accounts/repo-tests/src/accounts_repository_test_utils.rs (hunk 11 above)
Comment (rmn-boiko) [reply src/infra/accounts/repo-tests/src/accounts_repository_test_utils.rs:24]: In test we always were using `new_seeded_ed25519` method for AccountID which always return static value. But to test new logic with pregenerated AccountID with randomly generated Private part of did.

File: src/utils/random-strings/src/common.rs (hunk 17 above)
Comment (rmn-boiko) [reply src/utils/random-strings/src/common.rs:13]: Do we really want to duplicate 2 pretty similar methods. 
New method `get_random_string()` with `Alphanumeric` param works similar to old one

File: src/utils/crypto-utils/src/aes_gcm.rs (hunk 14 above)
Comment (rmn-boiko) [reply src/utils/crypto-utils/src/aes_gcm.rs:69]: Yeah, you're right. Changed

File: migrations/postgres/20250426151069_did_secret_keys_table.sql [hunk 18]
```diff
@@ -0,0 +1,13 @@
+CREATE TABLE dataset_did_secret_keys(
```
Comment (zaychenko-sergei) [inline migrations/postgres/20250426151069_did_secret_keys_table.sql:1]: All our migrations have separating comments. Please follow the format of other migrations.

File: migrations/sqlite/20240701092841_dataset_env_vars_table.sql [hunk 19]
```diff
@@ -4,7 +4,7 @@ CREATE TABLE dataset_env_vars(
     id VARCHAR(36) NOT NULL PRIMARY KEY,
     key VARCHAR(200) NOT NULL,
     value BLOB NOT NULL,
-    secret_nonce BLOKB,
+    secret_nonce BLOB,
```
[the remark this answers is not in the document]
Comment (zaychenko-sergei) [reply migrations/sqlite/20240701092841_dataset_env_vars_table.sql:7]: Yes, you can't modify existing migrations. You need to create a migration with new/old  column mapping, as changing the type will unlikely be possible directly in SQlite.

File: migrations/postgres/20250426151069_did_secret_keys_table.sql [hunk 20]
```diff
@@ -0,0 +1,13 @@
...
+   secret_key BYTEA NOT NULL,
+   owner_id VARCHAR(100) NOT NULL REFERENCES accounts(id) ON DELETE CASCADE
+);
+
+CREATE TABLE account_did_secret_keys(
+   account_id VARCHAR(100) NOT NULL REFERENCES accounts(id) ON DELETE CASCADE,
+   secret_nonce BYTEA NOT NULL,
+   secret_key BYTEA NOT NULL,
+   owner_id VARCHAR(100) NOT NULL REFERENCES accounts(id) ON DELETE CASCADE
+);
```
Comment (zaychenko-sergei) [inline migrations/postgres/20250426151069_did_secret_keys_table.sql:13]: What is the difference between `owner_id` and `account_id` in this context?

File: migrations/postgres/20250426151069_did_secret_keys_table.sql [hunk 21]
```diff
@@ -0,0 +1,13 @@
+CREATE TABLE dataset_did_secret_keys(
+   dataset_id VARCHAR(100) NOT NULL REFERENCES dataset_entries(dataset_id) ON DELETE CASCADE,
+   secret_nonce BYTEA NOT NULL,
+   secret_key BYTEA NOT NULL,
+   owner_id VARCHAR(100) NOT NULL REFERENCES accounts(id) ON DELETE CASCADE
```
Comment (zaychenko-sergei) [inline migrations/postgres/20250426151069_did_secret_keys_table.sql:5]: Why do we need `owner_id` here? We have `owner_id` in `DatasetEntry` already. Do you see any reason for denormalization from the beginning?

File: src/adapter/graphql/src/mutations/account_mut.rs [hunk 22]
```diff
@@ -57,6 +70,81 @@ impl AccountMut {
...
+                    field: err.account_field.to_string(),
+                }),
+            ),
+            Err(CreateAccountError::Internal(e)) => Err(e.into()),
+        }
+    }
+
+    /// Reset password for selected account. Allowed only for admin users
+    #[graphql(guard = "AdminGuard::new()")]
+    async fn modify_password(
```
Comment (zaychenko-sergei) [inline src/adapter/graphql/src/mutations/account_mut.rs:127]: Why the permission check is different from `create_account`? I don't think we are going to give Molecule an admin account for the whole system.

File: migrations/postgres/20250426151069_did_secret_keys_table.sql (hunk 21 above)
Comment (rmn-boiko) [reply migrations/postgres/20250426151069_did_secret_keys_table.sql:5]: Very soon we will add possibility to create datasets in one account from another. And to have here entity who creates it will be very useful.

File: src/adapter/graphql/src/utils.rs [hunk 23]
```diff
@@ -223,6 +224,27 @@ pub(crate) fn logged_account(ctx: &Context<'_>) -> bool {
...
+    let rebac_service = from_catalog_n!(ctx, dyn RebacService);
+
+    let account_properties = rebac_service
+        .get_account_properties(account_id)
+        .await
+        .int_err()?;
+
+    if account_properties.is_admin || account_properties.can_provision_accounts {
+        return Ok(());
+    }
```
Comment (zaychenko-sergei) [inline src/adapter/graphql/src/utils.rs:240]: Anti-pattern "feature envy" here.
This business logic should be encapsulated somewhere else in a domain-specific service rather than in GQL adapter.

File: migrations/postgres/20250426151069_did_secret_keys_table.sql (hunk 20 above)
Comment (rmn-boiko) [reply migrations/postgres/20250426151069_did_secret_keys_table.sql:13]: `owner_id` is Id of account who created this account, and `account_id` is Id of created account

File: src/adapter/graphql/src/mutations/account_mut.rs [hunk 24]
```diff
@@ -57,6 +70,81 @@ impl AccountMut {
...
+            Ok(created_account) => Ok(CreateAccountResult::Success(CreateAccountSuccess {
+                account: AccountView::from_account(created_account),
+            })),
+            Err(CreateAccountError::Duplicate(err)) => Ok(
+                CreateAccountResult::NonUniqueAccountField(AccountFieldNonUnique {
+                    field: err.account_field.to_string(),
+                }),
+            ),
+            Err(CreateAccountError::Internal(e)) => Err(e.into()),
+        }
```
Comment (zaychenko-sergei) [inline src/adapter/graphql/src/mutations/account_mut.rs:122]: Think of extracting a use case in `accounts` domain. This has nothing to do with GQL in particular.

File: src/adapter/graphql/src/mutations/account_mut.rs [hunk 25]
```diff
@@ -57,6 +70,81 @@ impl AccountMut {
...
+            .await
+        {
+            Ok(_) => Ok(ModifyPasswordResult::Success(
+                ModifyPasswordSuccess::default(),
+            )),
+            Err(ModifyPasswordError::AccountNotFound(_)) => Ok(
+                ModifyPasswordResult::AccountNotFound(AccountNotFound::default()),
+            ),
+            Err(ModifyPasswordError::Internal(e)) => Err(e.into()),
+        }
```
Comment (zaychenko-sergei) [inline src/adapter/graphql/src/mutations/account_mut.rs:145]: A bit smaller, but still smells like use case

File: src/domain/datasets/services/src/use_cases/create_dataset_from_snapshot_use_case_impl.rs [hunk 26]
```diff
@@ -86,6 +115,16 @@ impl CreateDatasetFromSnapshotUseCase for CreateDatasetFromSnapshotUseCaseImpl {
...
 
+        if let Some(did_secret_encryption_key) = &self.did_secret_encryption_key {
+            let dataset_did_secret_key =
+                DidSecretKey::try_new(&dataset_did.1, did_secret_encryption_key.expose_secret())
+                    .int_err()?;
+            self.dataset_did_secret_key_repo
+                .save_did_secret_key(&dataset_did.0, logged_account_id, &dataset_did_secret_key)
+                .await
+                .int_err()?;
+        }
```
Comment (zaychenko-sergei) [inline src/domain/datasets/services/src/use_cases/create_dataset_from_snapshot_use_case_impl.rs:126]: Don't we need the same functionality in "create dataset" w/o snapshot use case? I think it should be placed in the shared create helper that we have, and invoked from both use cases.

File: src/infra/accounts/repo-tests/src/password_hash_repository_test_suite.rs (hunk 12 above)
Comment (zaychenko-sergei) [reply src/infra/accounts/repo-tests/src/password_hash_repository_test_suite.rs:100]: The value type you suggested (nutype-based) would handle those better.

File: src/adapter/graphql/src/utils.rs (hunk 23 above)
Comment (rmn-boiko) [reply src/adapter/graphql/src/utils.rs:240]: I would disagree with that. This is purely responsibility of Oso rebac for gql methods. I don't see this logic on domain level. 
Probably confusing thing that I call directly `rebac_service` instead Oso. The reason I did it that recently we migrate caching level to the `rebac_service` level. I can agree to change the calling to Oso service but can't see any other place for this logic

File: src/domain/datasets/services/src/use_cases/create_dataset_from_snapshot_use_case_impl.rs (hunk 26 above)
Comment (rmn-boiko) [reply src/domain/datasets/services/src/use_cases/create_dataset_from_snapshot_use_case_impl.rs:126]: It is a good and tricky question. Goal of epic to store all Did keys created via node. But `create_dataset_use_case` is used only in transfer protocol. And seems we don't need to store dids for pushed/pulled dataset

File: src/adapter/graphql/src/mutations/account_mut.rs (hunk 24 above)
Comment (rmn-boiko) [reply src/adapter/graphql/src/mutations/account_mut.rs:122]: Agree, moved.

File: migrations/postgres/20250426151069_did_secret_keys_table.sql (hunk 20 above)
Comment (zaychenko-sergei) [reply migrations/postgres/20250426151069_did_secret_keys_table.sql:13]: Then `owner_id` => `creator_id`.

File: src/adapter/graphql/src/mutations/account_mut.rs (hunk 22 above)
Comment (zaychenko-sergei) [reply src/adapter/graphql/src/mutations/account_mut.rs:127]: Who wil be modifying the passwords, us for all customers?

File: src/domain/datasets/services/src/use_cases/create_dataset_from_snapshot_use_case_impl.rs (hunk 26 above)
Comment (zaychenko-sergei) [reply src/domain/datasets/services/src/use_cases/create_dataset_from_snapshot_use_case_impl.rs:126]: That is weird. We allow creating datasets in UI without specifying any snapshot.

File: src/adapter/graphql/src/utils.rs (hunk 23 above)
Comment (zaychenko-sergei) [reply src/adapter/graphql/src/utils.rs:240]: I expect something like "account facade", so that I can simply ask:

```
let account_facade = from_catalog_n!(ctx, dyn RebacAccountFacade);
account_facade.can_provision_accounts(account_id).await
```

While checking for admins or knowing this data exists in account properties - is a ReBAC domain implementation details.

File: src/domain/datasets/services/src/use_cases/create_dataset_from_snapshot_use_case_impl.rs (hunk 26 above)
Comment (rmn-boiko) [reply src/domain/datasets/services/src/use_cases/create_dataset_from_snapshot_use_case_impl.rs:126]: YEah, but under the hood we still calling `create_from_snapshot_impl` with empty snapshot

File: examples/archive/commercial-fishing/.kamuconfig [hunk 27]
```diff
@@ -4,7 +4,7 @@ content:
   users:
     predefined:
       - accountName: kamu
-        isAdmin: true
+        properties: [isAdmin]
```
[the remark this answers is not in the document]
Comment (rmn-boiko) [reply examples/archive/commercial-fishing/.kamuconfig:7]: Yeah, I wanted to do it but a bit scared of additional migration. But agree it would fit better

File: src/domain/accounts/domain/src/entities/password.rs [hunk 28]
```diff
@@ -0,0 +1,55 @@
...
+// by the Apache License, Version 2.0.
+
+use nutype::nutype;
+use thiserror::Error;
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[nutype(
+    sanitize(trim),
+    validate(with = validate_password, error = PasswordValidationError),
```
Comment (zaychenko-sergei) [inline src/domain/accounts/domain/src/entities/password.rs:17]: There is an existing length validation in `nutype`: `len_char_min`. The non-ascii part could be expressed via Regex, but custom is fine there too.

File: src/domain/accounts/services/src/use_cases/create_account_use_case_impl.rs [hunk 29]
```diff
@@ -0,0 +1,62 @@
...
+                format!(
+                    "{}+{}@{}",
+                    parent_account_parts.0, account_name, parent_account_parts.1
+                )
+            } else {
+                format!("{account_name}@example.com")
+            };
+
+            Email::parse(&email_str).int_err()?
+        });
```
Comment (zaychenko-sergei) [inline src/domain/accounts/services/src/use_cases/create_account_use_case_impl.rs:46]: Let's encapsulate the validation inside `Email` type.

File: src/domain/datasets/services/src/use_cases/create_dataset_from_snapshot_use_case_impl.rs [hunk 30]
```diff
@@ -19,20 +19,55 @@ use kamu_datasets::{
...
+    pub fn new(
+        current_account_subject: Arc<CurrentAccountSubject>,
+        system_time_source: Arc<dyn SystemTimeSource>,
+        did_generator: Arc<dyn DidGenerator>,
+        dataset_registry: Arc<dyn DatasetRegistry>,
+        create_helper: Arc<CreateDatasetUseCaseHelper>,
+        did_secret_encryption_config: Arc<DidSecretEncryptionConfig>,
+        did_secret_key_repo: Arc<dyn DidSecretKeyRepository>,
+    ) -> Self {
+        Self {
```
Comment (zaychenko-sergei) [inline src/domain/datasets/services/src/use_cases/create_dataset_from_snapshot_use_case_impl.rs:58]: I still don't understand why we are extending "create from snapshot" and not extending "create" similarly.

File: migrations/postgres/20250426151069_did_secret_keys_table.sql [hunk 31]
```diff
@@ -0,0 +1,18 @@
...
+CREATE TYPE did_entity_type AS ENUM ('account', 'dataset');
+
+/* ------------------------------ */
+
+CREATE TABLE did_secret_keys(
+   entity_type  did_entity_type NOT NULL,
+   entity_id    VARCHAR(100) NOT NULL,
+   secret_nonce BYTEA NOT NULL,
+   secret_key   BYTEA NOT NULL,
+   creator_id   VARCHAR(100) NOT NULL REFERENCES accounts(id) ON DELETE CASCADE
```
Comment (zaychenko-sergei) [inline migrations/postgres/20250426151069_did_secret_keys_table.sql:12]: Doubts here:
- the concepts of "Account" and "DidSecret" feel closely related (it feels like account exclusively owns 0..N secrets), so the cascade delete looks adequate
- it feels mismatching to put them in separate DDD domains, as separate domains should not use something like cascade delete and should be losely connected via outbox messages.

My gut feeling is that there is no motivation to place "secrets" in separate domain, it naturally belongs to "accounts" domain. If we are keeping these separate for some motivating reason, then we should not have this database-level coupling.

File: src/domain/accounts/domain/src/entities/predefined_accounts_config.rs (hunk 8 above)
Comment (s373r) [reply src/domain/accounts/domain/src/entities/predefined_accounts_config.rs:141]: Yes, I confused you by mentioning the account name -- I apologize

I wanted to emphasize that if the method is called `set_properties()`, the expected behavior is setter behavior, and in this case we have an update.

And it's not quite clear to me how to delete properties, for example -- hence the suggestion to make a more explicit (and simple) API

File: examples/archive/commercial-fishing/.kamuconfig (hunk 27 above)
[the remark this answers is not in the document]
Comment (s373r) [reply examples/archive/commercial-fishing/.kamuconfig:7]: It's a little unclear to me -- there's a reaction, but no change

File: src/domain/datasets/domain/src/repos/dataset_did_secret_key_repository.rs (hunk 6 above)
Comment (s373r) [reply src/domain/datasets/domain/src/repos/dataset_did_secret_key_repository.rs]: > In DimaP's wallet auth work it looks like we will adopt did:ethr:<public key hash> or did:pkh for identifying EVM accounts.

Cross-posting: `did:pkh` has been selected

File: src/utils/crypto-utils/src/aes_gcm.rs (hunk 14 above)
Comment (s373r) [reply src/utils/crypto-utils/src/aes_gcm.rs]: `Display` implementation can also be hidden via `#[error("AES-GCM error")`

File: src/utils/random-strings/src/common.rs (hunk 17 above)
Comment (s373r) [reply src/utils/random-strings/src/common.rs:13]: These methods, different semantically. For example, here are the random names that Docker generates:
image

That is, they are readable and can be easily reproduced.

In my opinion, we can keep `get_random_name()`, which will just do the `get_random_string()` call internally, so as not to change all the usage locations -- and then replace the `get_random_name()` implementation with something more appropriate whenever we want it

File: src/utils/crypto-utils/src/aes_gcm.rs (hunk 14 above)
Comment (rmn-boiko) [reply src/utils/crypto-utils/src/aes_gcm.rs]: Changed, thanks

File: src/domain/accounts/services/src/services/predefined_accounts_registrator.rs [hunk 32]
```diff
@@ -55,8 +55,26 @@ impl PredefinedAccountsRegistrator {
         account_id: &AccountID,
         account_config: &AccountConfig,
     ) -> Result<(), InternalError> {
+        let default_account_properties = vec![
+            AccountPropertyName::is_admin(false),
+            AccountPropertyName::can_provision_accounts(false),
+        ];
```
[the remark this answers is not in the document]
Comment (rmn-boiko) [reply src/domain/accounts/services/src/services/predefined_accounts_registrator.rs:61]: Used this value

File: migrations/postgres/20250426151069_did_secret_keys_table.sql (hunk 31 above)
Comment (rmn-boiko) [reply migrations/postgres/20250426151069_did_secret_keys_table.sql:12]: Changed

File: src/domain/accounts/domain/src/entities/password.rs (hunk 28 above)
Comment (rmn-boiko) [reply src/domain/accounts/domain/src/entities/password.rs:17]: Yeah, just don't want to use here regex, and use single custom validation

File: src/utils/random-strings/src/common.rs (hunk 17 above)
Comment (rmn-boiko) [reply src/utils/random-strings/src/common.rs:13]: Agree with that, Changed

File: src/domain/accounts/domain/src/entities/predefined_accounts_config.rs (hunk 8 above)
Comment (rmn-boiko) [reply src/domain/accounts/domain/src/entities/predefined_accounts_config.rs:141]: Such as it is mainly used in testing I changed to rewrite properties

File: src/domain/datasets/services/src/use_cases/create_dataset_from_snapshot_use_case_impl.rs (hunk 30 above)
Comment (rmn-boiko) [reply src/domain/datasets/services/src/use_cases/create_dataset_from_snapshot_use_case_impl.rs:58]: `CreateDatasetUseCase` is used only in Smtp protocol. And during this flow we already expect generated seed block with dataset Did without its private part. So we can't implement this logic here and moreover in current situation we don't have to

File: src/adapter/graphql/src/mutations/account_mut.rs (hunk 25 above)
Comment (rmn-boiko) [reply src/adapter/graphql/src/mutations/account_mut.rs]: I'm not sure 1 line deserve to be a separate usecase

**Review by zaychenko-sergei — APPROVED** [review]
